src/rng.rs

```diff
@@ -93,40 +94,75 @@ pub trait Rng: RngCore {
-    /// Generate a random value in the range [`low`, `high`), i.e. inclusive of
-    /// `low` and exclusive of `high`.
+/*
+macro_rules! impl_distrib_range_to {
+    ($($ty:ty),*) => {
+        $(
+            impl Distribution<$ty> for RangeTo<$ty> {
+                fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> $ty {
+                    <$ty as SampleUniform>::Sampler::sample_single(0, self.end, rng)
+                }
+            }
+            impl Distribution<$ty> for RangeToInclusive<$ty> {
+                fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> $ty {
+                    <$ty as SampleUniform>::Sampler::sample_single_inclusive(0, self.end, rng)
+                }
+            }
+        )*
+    }
+}
+
+impl_distrib_range_to!(usize, u8, u16, u32, u64);
+#[cfg(not(target_os = "emscripten"))]
+impl_distrib_range_to!(u128);
+*/
+
+    /// Generate a random value in the given range.
     ///
     /// This function is optimised for the case that only a single sample is
     /// made from the given range. See also the [`Uniform`] distribution
     /// type which may be faster if sampling from the same range repeatedly.
     ///
-    /// Note that `Rng::gen_range(low, high)` is equivalent to
-    /// `Rng::sample(low..high)`. If you need a random value from the range
-    /// including `high`, you can use `Rng::sample(low..=high)`.
+    /// Only `gen_range(low..high)` and `gen_range(low..=high)` are supported.
     ///
     /// # Panics
     ///
-    /// Panics if `low >= high`.
+    /// Panics if the range is not `low..high` or `low..=high`.
     ///
     /// # Example
     ///
     /// ```
     /// use rand::{thread_rng, Rng};
     ///
     /// let mut rng = thread_rng();
-    /// let n: u32 = rng.gen_range(0, 10);
+    ///
+    /// // Exclusive range
+    /// let n: u32 = rng.gen_range(0..10);
     /// println!("{}", n);
-    /// let m: f64 = rng.gen_range(-40.0f64, 1.3e5f64);
+    /// let m: f64 = rng.gen_range(-40.0f64..1.3e5f64);
     /// println!("{}", m);
+    ///
+    /// // Inclusive range
+    /// let n: u32 = rng.gen_range(0..=10);
+    /// println!("{}", n);
     /// ```
     ///
     /// [`Uniform`]: distributions::uniform::Uniform
-    fn gen_range<T: SampleUniform, B1, B2>(&mut self, low: B1, high: B2) -> T
+    fn gen_range<T, R>(&mut self, range: R) -> T
     where
-        B1: SampleBorrow<T> + Sized,
-        B2: SampleBorrow<T> + Sized,
+        T: SampleUniform,
+        R: RangeBounds<T>
     {
-        T::Sampler::sample_single(low, high, self)
+        use core::ops::Bound;
+        if let Bound::Included(low) = range.start_bound() {
+            match range.end_bound() {
+                Bound::Excluded(high) => T::Sampler::sample_single(low, high, self),
+                Bound::Included(high) => T::Sampler::sample_single_inclusive(low, high, self),
+                Bound::Unbounded => panic!("invalid upper bound"),
+            }
+        } else {
+            panic!("invalid lower bound");
+        }
```

Comment:
We shouldn't have to use run-time look-ups and panics. Can we do something like: `where R: Into<Uniform<T>>`?

No, really we need a specific trait for "sample single":
```rust
trait SampleSingle<X> {
    fn sample_single(self) -> X;
}
impl<X: SampleUniform> SampleSingle<X> for Range<X> { .. }
impl<X: SampleUniform> SampleSingle<X> for RangeInclusive<X> { .. }
```

Reply:
Yes, we would have to define such a trait if we want to avoid the lookup. Alternatively, we could inline `gen_range` and hope the lookup gets optimized away. What do you prefer?

Reply:
I added an implementation using static dispatch in the last commit. It required a new public trait that we may want to hide with `#[doc(hidden)]` to avoid users implementing it to add support for other range types.

Reply:
What do you think @newpavlov? Overall it seems more useful to expose the trait (sometimes users do want to extend generic APIs).